// minecraft-server-util/dist/structure/UDPClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dgram_1 = __importDefault(require("dgram"));
const events_1 = require("events");
const util_1 = require("util");
const varint_1 = require("../util/varint");
const encoder = new util_1.TextEncoder();
const decoder = new util_1.TextDecoder('utf-8');
class UDPClient extends events_1.EventEmitter {
    constructor(host, port) {
        super();
        this.data = Buffer.alloc(0);
        this.host = host;
        this.port = port;
        this.socket = dgram_1.default.createSocket('udp4');
        this.socket.on('message', (data) => {
            this.data = Buffer.concat([this.data, data]);
            this.emit('data');
        });
    }
    readByte() {
        return this.readUInt8();
    }
    writeByte(value) {
        this.writeUInt8(value);
    }
    readBytes(length) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(length);
            const value = this.data.slice(0, length);
            this.data = this.data.slice(length);
            return value;
        });
    }
    writeBytes(data) {
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt8() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(1);
            const value = this.data.readUInt8(0);
            this.data = this.data.slice(1);
            return value;
        });
    }
    writeUInt8(value) {
        const data = Buffer.alloc(1);
        data.writeUInt8(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt8() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(1);
            const value = this.data.readInt8(0);
            this.data = this.data.slice(1);
            return value;
        });
    }
    writeInt8(value) {
        const data = Buffer.alloc(1);
        data.writeInt8(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt16BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(2);
            const value = this.data.readUInt16BE(0);
            this.data = this.data.slice(2);
            return value;
        });
    }
    writeUInt16BE(value) {
        const data = Buffer.alloc(2);
        data.writeUInt16BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt16BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(2);
            const value = this.data.readInt16BE(0);
            this.data = this.data.slice(2);
            return value;
        });
    }
    writeInt16BE(value) {
        const data = Buffer.alloc(2);
        data.writeInt16BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt16LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(2);
            const value = this.data.readUInt16LE(0);
            this.data = this.data.slice(2);
            return value;
        });
    }
    writeUInt16LE(value) {
        const data = Buffer.alloc(2);
        data.writeUInt16LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt16LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(2);
            const value = this.data.readInt16LE(0);
            this.data = this.data.slice(2);
            return value;
        });
    }
    writeInt16LE(value) {
        const data = Buffer.alloc(2);
        data.writeInt16LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt32BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readUInt32BE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeUInt32BE(value) {
        const data = Buffer.alloc(4);
        data.writeUInt32BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt32BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readInt32BE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeInt32BE(value) {
        const data = Buffer.alloc(4);
        data.writeInt32BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt32LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readUInt32LE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeUInt32LE(value) {
        const data = Buffer.alloc(4);
        data.writeUInt32LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt32LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readInt32LE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeInt32LE(value) {
        const data = Buffer.alloc(4);
        data.writeInt32LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt64BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readBigUInt64BE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeUInt64BE(value) {
        const data = Buffer.alloc(8);
        data.writeBigUInt64BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt64BE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readBigInt64BE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeInt64BE(value) {
        const data = Buffer.alloc(8);
        data.writeBigInt64BE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readUInt64LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readBigUInt64LE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeUInt64LE(value) {
        const data = Buffer.alloc(8);
        data.writeBigUInt64LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readInt64LE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readBigInt64LE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeInt64LE(value) {
        const data = Buffer.alloc(8);
        data.writeBigInt64LE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readFloatBE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readFloatBE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeFloatBE(value) {
        const data = Buffer.alloc(4);
        data.writeFloatBE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readFloatLE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(4);
            const value = this.data.readFloatLE(0);
            this.data = this.data.slice(4);
            return value;
        });
    }
    writeFloatLE(value) {
        const data = Buffer.alloc(4);
        data.writeFloatLE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readDoubleBE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readDoubleBE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeDoubleBE(value) {
        const data = Buffer.alloc(8);
        data.writeDoubleBE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readDoubleLE() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ensureBufferedData(8);
            const value = this.data.readDoubleLE(0);
            this.data = this.data.slice(8);
            return value;
        });
    }
    writeDoubleLE(value) {
        const data = Buffer.alloc(8);
        data.writeDoubleLE(value);
        this.data = Buffer.concat([this.data, data]);
    }
    readVarInt() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, varint_1.readVarInt)(() => this.readByte());
        });
    }
    writeVarInt(value) {
        this.writeBytes((0, varint_1.writeVarInt)(value));
    }
    readString(length) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.readBytes(length);
            return decoder.decode(data);
        });
    }
    writeString(value) {
        this.writeBytes(encoder.encode(value));
    }
    readStringVarInt() {
        return __awaiter(this, void 0, void 0, function* () {
            const length = yield this.readVarInt();
            const data = yield this.readBytes(length);
            return Array.from(data).map((point) => String.fromCodePoint(point)).join('');
        });
    }
    writeStringVarInt(value) {
        const data = encoder.encode(value);
        this.writeVarInt(data.byteLength);
        this.writeBytes(data);
    }
    readStringNT() {
        return __awaiter(this, void 0, void 0, function* () {
            let buf = Buffer.alloc(0);
            let value;
            while ((value = yield this.readByte()) !== 0x00) {
                buf = Buffer.concat([buf, Buffer.from([value])]);
            }
            return Array.from(buf).map((point) => String.fromCodePoint(point)).join('');
        });
    }
    readStringNTFollowedBy(suffixes) {
        return __awaiter(this, void 0, void 0, function* () {
            let buf = Buffer.alloc(0);
            // eslint-disable-next-line no-constant-condition
            while (true) {
                const value = yield this.readByte();
                if (value === 0x00 && (yield this.checkUpcomingData(suffixes))) {
                    break;
                }
                buf = Buffer.concat([buf, Buffer.from([value])]);
            }
            return Array.from(buf).map((point) => String.fromCodePoint(point)).join('');
        });
    }
    checkUpcomingData(suffixes) {
        return __awaiter(this, void 0, void 0, function* () {
            let i = 0;
            while (suffixes.length) {
                yield this.ensureBufferedData(i + 1);
                const remaining = [];
                for (const suffix of suffixes) {
                    if (this.data[i] === suffix[i]) {
                        if (i === suffix.length - 1) {
                            return suffix;
                        }
                        remaining.push(suffix);
                    }
                }
                suffixes = remaining;
                i++;
            }
            return null;
        });
    }
    writeStringNT(value) {
        const data = encoder.encode(value);
        this.writeBytes(data);
        this.writeByte(0x00);
    }
    writeStringBytes(value) {
        this.writeBytes(encoder.encode(value));
    }
    flush(prefixLength = true) {
        if (!this.socket)
            return Promise.resolve();
        return new Promise((resolve, reject) => {
            let buf = this.data;
            if (prefixLength) {
                buf = Buffer.concat([(0, varint_1.writeVarInt)(buf.byteLength), buf]);
            }
            this.socket.send(buf, 0, buf.byteLength, this.port, this.host, (error) => {
                if (error)
                    return reject(error);
                resolve();
            });
            this.data = Buffer.alloc(0);
        });
    }
    close() {
        var _a;
        try {
            (_a = this.socket) === null || _a === void 0 ? void 0 : _a.close();
        }
        catch (_b) {
            // Ignore
        }
    }
    ensureBufferedData(byteLength) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.data.byteLength >= byteLength)
                return Promise.resolve();
            return this._waitForData(byteLength);
        });
    }
    _waitForData(byteLength = 1) {
        return new Promise((resolve, reject) => {
            const dataHandler = () => {
                if (this.data.byteLength >= byteLength) {
                    this.removeListener('data', dataHandler);
                    this.socket.removeListener('error', errorHandler);
                    resolve();
                }
            };
            const errorHandler = (error) => {
                this.removeListener('data', dataHandler);
                this.socket.removeListener('error', errorHandler);
                reject(error);
            };
            this.once('data', () => dataHandler());
            this.socket.on('error', (error) => errorHandler(error));
        });
    }
    hasRemainingData() {
        return this.data.byteLength > 0;
    }
}
exports.default = UDPClient;
